basic rbac support

jwfang opened the pull request:
replace #1351 

some of the works are from @rajiteh  

changes summary:

1. system components use their own certs, identified as user/group document here:
    https://kubernetes.io/docs/admin/authorization/rbac/#default-roles-and-role-bindings

    changed components:  kubelet/kube-proxy/controller-manager/scheduler

2. addon run as ServiceAccount and changed related ClusterRole/ClusterRoleBinding,
    patch existing Binding if necessary.

   changed addons:  kube-dns/helm

3. add an *authorization_modes* var to control authorization modes

jwfang commented:
messed it up, create a new one. sorry

bradbeam commented:
Is this supposed to replace #1351? Seems like there are a number of missing files.

jwfang commented:
@bradbeam yes, this replace #1351 . forgot -a, added.

bradbeam commented:
Two more asks --
I just merged in an update to kube-dns and it looks like that caused some conflicts. Would you be able to address?
Could you add a test case for RBAC in ci?

jwfang commented:
@bradbeam rebased on latest master

i am not sure how to add unit tests. should i add RBAC for all existing test ? there seem to be a lots combinations.

seungkyua commented:
I think using cluseterrolebinding for serviceaccount and node is more simple.
There is a apiserver_custom_flags in roles/kubernetes/master/defaults/main.yml file.
If you want, put the "--authorization-mode=RBAC" parameters there.

jwfang commented:
@seungkyua thanks for replying. can you explain it in more detail ? i didn't catch you.

> I think using cluseterrolebinding for serviceaccount and node is more simple.
i am just following the RBAC docs for builtin user/group/rolebinding, i am not sure what you mean
here. :)

> There is a apiserver_custom_flags in roles/kubernetes/master/defaults/main.yml file.
> If you want, put the "--authorization-mode=RBAC" parameters there.
we added the *authorization-modes* to control the whole RBAC things, *apiserver_custom_flags * is only one of them. you can find the detailed commit comment in #1350 .

hzamani commented:
We have just tested this patch, everything was OK except kube-dns. To fix it we had to:
- upgrade kube-dns version to 1.14.4
- add kube-dns [config map](https://github.com/kubernetes/kubernetes/blob/master/cluster/addons/dns/kubedns-cm.yaml)
- add configmaps to system:kube-dns cluster role resources

jwfang commented:
@hzamani kube-dns is recently updated, i haven't got the time to update and test against it.
could you make a PR and i will merge it.

@bradbeam could you give me some hint on how to write ci test so we can push this forward?

@rajiteh unfortunately, i didn't have enough time for this. if you or some one want to do this, could take over from here.

bradbeam commented:
For CI
- Add in a new variable to control authorization modes
I think https://github.com/kubernetes-incubator/kubespray/blob/master/.gitlab-ci.yml#L64 we could add in `AUTHORIZATION_MODES='{ "authorization_modes": ["AlwaysAllow"] }'` ( to set the default )
- Add in the new variable to the ansible command
https://github.com/kubernetes-incubator/kubespray/blob/master/.gitlab-ci.yml#L133 something like `-e authorization_modes=${AUTHORIZATION_MODES}`
- Add in a new set of variables for rbac test https://github.com/kubernetes-incubator/kubespray/blob/master/.gitlab-ci.yml#L371
```
.ubuntu_calico_rbac_variables: &ubuntu_calico_rbac_variables
# stage: deploy-gce-special
  AUTHORIZATION_MODES: '{ "authorization_modes":  [ "RBAC" ] }'
  KUBE_NETWORK_PLUGIN: calico
  CLOUD_IMAGE: ubuntu-1604-xenial
  CLOUD_REGION: europe-west1-b
  CLUSTER_MODE: separate
  STARTUP_SCRIPT: ""
```
-  Actual test https://github.com/kubernetes-incubator/kubespray/blob/master/.gitlab-ci.yml#L596
```
coreos-alpha-weave-ha:
ubuntu-calico-rbac-sep:
  stage: deploy-gce-special
  <<: *job
  <<: *gce
  variables:
    <<: *gce_variables
    <<: *ubuntu_calico_rbac_variables
  when: manual
  except: ['triggers']
  only: ['master', /^pr-.*$/]
```

i _think_ that should do it

jwfang commented:
@bradbeam thanks for detailed guide, i can just copy & paste from it. test added.

@hzamani with d5516a4ca9c42cc6cfb4c1aef692e0ef323abfcc kube-dns works without configmap.
and create a ConfigMap *kube-dns* in *kube-system*, kube-dns will pick up the new config as expected.
so i thin we don't need to do anything regarding ConfigMap.

rajiteh commented:
> @rajiteh unfortunately, i didn't have enough time for this. if you or some one want to do this, could take over from here.

@jwfang could you clarify? i can help out

jwfang commented:
@rajiteh i have some other tasks from work, and may not have enough time to keep track on this.
i guess you can create a new PR from this. when you've done that, i will close this one.

jwfang commented:
@bradbeam anything we can do to make the tested and get this merged ? thanks.

bradbeam commented:
ci check this

hzamani commented:
Calico controller requires RBAC rules to access pods and namespaces and create network polices.

jwfang commented:
@hzamani thanks for your advice. 

i am not familiary with calico, found https://github.com/projectcalico/calico/issues/402 and https://github.com/projectcalico/calico/blob/master/master/getting-started/kubernetes/installation/rbac.yaml , but not sure if they works.

anyway, PR are welcome.

@bradbeam i rebase for the lastest master, should we add some RBAC & calico checks for now
and do RBAC for calico later ? and change the test for non-calico-RBAC ?
if this is acceptable, i will submit another commit.

hzamani commented:
@jwfang The `rbac.yaml` from project calico is just what's needed.

jwfang commented:
@hzamani thanks.
i am new with calico. there seem to be serval components for calico, and i never get calico works in my environment ( i am using flannel).
we have *calico* and *calico-rr* roles, and there seems to be lots of features/components for calico itself.
can you give me some guide on where should i put the *rbac.yaml* from calico ?

rsmitty commented:
I'm a bit unclear on where we stand on this PR. What's needed to get to a working state for this? Anyone tried it yet?

hzamani commented:
@jwfang I found it confusing to not have calico in kubernetes-apps/network_plugin, and it's a good idea to place calico_node daemon set there.

@rsmitty I tried it before calico related stuff and except for calico (which I had to add RBAC rules by hand to make it work) everything was OK.

seungkyua commented:
@jwfang  sorry for late comment.

I think this is so complicated.
To use a rbac, as I know, we just make a ClusterRoleBinding about "admin users", "k8s nodes" and "default service account in {{system-namespaces}}" reference to ClusterRole about "cluster-admin".
It's good worked to my environment.

jwfang commented:
@rsmitty @hzamani 
currently i have trouble to get calico-RBAC working. 
from https://github.com/projectcalico/calico/blob/master/master/getting-started/kubernetes/installation/rbac.yaml, we need do two parts: *calico-policy-controller* and *calico-node*. i have do the *calico-policy-controller* part, but not sure how to do the *calico-node* part.

after a second thought, i guess we may not need to change *calico-node* to run as DamonSet. instead, we can just add certs for the current systemd service. i found api-server related stuff in https://github.com/kubernetes-incubator/kubespray/blob/992023288f4a42ce6960d611dc92c8ee798efd81/roles/network_plugin/calico/templates/cni-calico.conf.j2#L24, i think that's it.

i am new to cni/calico and not work full-time on this, so may not finish the calico stuff soon unless there is some guide/help.

@rsmitty if the other parts are fine, should we finish this PR and leave calico-RBAC to another PR ?

bradbeam commented:
@jwfang 
What network plugin have you been testing with? I think I'm fine with updating the CI tests with whatever plugin is working so we can get a successful CI run. 

If you have time, we've got #1271 logged to track the necessary work to uplift calico, but I'm not sure I want that to be a blocker for this PR. Would you be able to update the docs to reflect that RBAC is in experimental phase and note that we currently do not support RBAC with either calico or vault. 

I think once we get CI passing, I'm fine with getting this merged in.

hzamani commented:
@bradbeam we have successfully deployed Kubernetes using this patch. Thanks to @jwfang.

seungkyua commented:
@jwfang @bradbeam 
I suggest the rbac mode.
As I know, There are two ways to implement.

1. Using cluster-admin
    Only default serviceaccount in kube-system, admin user (root, kube) in k8s-cluster.yml and kube-node-{{ hostname }} user are have cluster-admin role.
    All of them are system user. So I think It's not a risk about security.

2. Using each user and group is binding to it's role.
    Clusterrolebinding                                 Name                                                Type
    --------------------------------------------------------------------------------
    system:kube-controller-manager  -->  system:kube-controller-manager  (User)
    system:kube-scheduler                  -->  system:kube-scheduler                  (User)
    system:node-proxier                       -->  system:kube-proxy                         (User)
    system:node                                    -->  system:nodes                                  (Group)
    system:kube-dns                            -->  kube-dns (only kube-system)        (ServiceAccount)
    cluster-admin   (tiller, calico etc)   --> default   (only kube-system)           (ServiceAccount)
    cluster-admin                                  --> root, kube                                         (Admin user)

I think we don't need to make a kube-config.yml and new cert name.

jwfang commented:
@bradbeam i am using flannel in my test environment.
i would like to do additonal calico/RBAC related works, but currently i think i am not qualified since i am not familiar with calico/cni and it seems would take a while to learn from scratch. 
all the cni modes are in my to-learn list, but i can't make any promise for the time frame since i am doing this part-time. 

for this PR, i did the following additional changes to make it mergable in the current state:
1. revert all calico related changes,
2. update docs say calico & vault not supported.
3. change gce test to use *flannel* instead of *calico*
4. add *authorization_modes* vars to all *cluster.yml* deployment

jwfang commented:
@seungkyua 
i am not sure i understanding you corrently and completely.

currently the user/role/rolebinding are take directly from the official k8s RBAC docs. i just try to stick with
the docs without taking short-cuts. :)

the reason for multiple *kube-config* is: a single *kube-config* file can have only one *current-context*, i could find a way to share it between different components.

bradbeam commented:
Thanks, kicking off the rbac ci test.

jwfang commented:
@bradbeam i see *delete gce instances* failed from ci, i guess that not related to this PR ?

bradbeam commented:
I'm curious if the failure is actually a bit above that with running the netchecker test

```
TASK [debug] *******************************************************************
task path: /builds/kargo-ci/kubernetes-incubator__kubespray/tests/testcases/040_check-network-adv.yml:70
Wednesday 12 July 2017  17:05:27 +0000 (0:00:00.425)       0:00:01.880 ******** 
An exception occurred during task execution. To see the full traceback, use -vvv. The error was: ValueError: No JSON object could be decoded
fatal: [k8s-9725808-22224346-2]: FAILED! => {"failed": true, "msg": "Unexpected failure during module execution.", "stdout": ""}
```
In looking at the netchecker deployment, I don't think there's anything that _should_ break when rbac is enabled. Would you be able to deploy netchecker to your environment and run through that testcases/040_check-network-adv.yml and see if you get different results?

jwfang commented:
@bradbeam sorry for the late, fixed the *040_check-network-adv.yml* test and rebased.
it turns out netchecker-server needs *list pods* privilege.

could you rerun the test for me ？

jwfang commented:
@bradbeam sorry, commit my local changes for #1381. fix it in 3d87f23

jwfang commented:
@bradbeam hi, what's needed for this PR to get merged ? seems the tests passed.

whereismyjetpack commented:
I don't recall the issue, but I was not able to "upgrade" an existing cluster when I tried.

jwfang commented:
@Starefossen no, i didn't test the upgrade process much. :(
i guess for RBAC -> RBAC upgrade, it will works as other parts of kubespray;
for no-RBAC -> RBAC, it would be a two-steps process: no-RBAC -> RBAC+AlwaysAllow -> RBAC-only, and there should be problems as noted by @whereismyjetpack.

currently i am hoping to get this merged as it's for the following reason:
1. i have not enough time for this and i am not work full-time on this, as noted above;
2. it's already a bit complex for this PR;
3.  other no-trival tasks, such as calico support, upgrade etc, could be done by others based on this.

whereismyjetpack commented:
I'm curious what problems with calico? I'm running a merged in copy of this PR with calico and it _ssems_ to work fine.

jwfang commented:
@whereismyjetpack see https://github.com/projectcalico/calico/blob/master/master/getting-started/kubernetes/installation/rbac.yaml and https://github.com/kubernetes-incubator/kubespray/issues/1271, i **guess** calico also need proper role/rolebinding.

PS: i am not familiar with calico and never get calico works in my environment (maybe due to environment issue which i didn't figure out yet). 
my original test for RBAC was using calico, and it failed in the CI test. so i switch to flannel for the CI in https://github.com/kubernetes-incubator/kubespray/pull/1382/commits/d24520161416cc586cfe0968fe66afc4722c4c0b

whereismyjetpack commented:
Thanks for that information, in my setup calico is seemingly deployed as a systemd unit file, and not a deployment within kube. i'm not sure under which conditions it chooses one over the other, or maybe that PR hasn't made it in yet? I don't want to hijack this PR with calico related convos, but wanted to thank you for replying

rsmitty commented:
Okay, so it looks like we're passing CI with this PR. Awesome! I want to verify my understanding a bit. The way it's setup right now, the RBAC options actually won't be active by default, right? Asking, because I don't want to affect existing clusters with this PR if possible. I also would love to see some research on the upgrade path with this. How we might support adding RBAC to an existing cluster.

Other than that, I think I'm cool with merging this. We can follow it up with any other necessary changes.

whereismyjetpack commented:
the `authorization_modes` variable needs to be set, it defaults to an empty list. leaving it unset will setup the cluster as it does today.

jwfang commented:
@rsmitty as @whereismyjetpack pointed out, it default with no-RBAC and keeps the current behaviour.

for upgrade from non-RBAC to RBAC, unfortunately i haven't got enough time to work on. :(

whereismyjetpack commented:
where are we with this PR? anything i can do to help?

joosthof commented:
Really need this PR for our clusters. Anything I can do to get this PR merged? Thanks!

jwfang commented:
cc @bradbeam @rsmitty

@whereismyjetpack @joisthof i think you can provide feedback, such as successful/failure/problem encountered etc.

anyone willing to take over this is welcome. :)

bradbeam commented:
Proceeding with merge. Will follow up with additional fixes as needed.

bradbeam commented:
@jwfang  thanks for this awesome pr, appreciate your patience
